---
title: "R Notebook"
output: html_notebook
---
NOTE: 100 genes
each has 200 mRNAs
same sequence for all genes (gene length = 493 codons, median length for 4839 genes)
same initiation probability for all genes = the lowest among 4839 genes
100 decay rates for 100 genes, which are the 1st-100th percentile of the Neymotin decay rates for 4839 genes (except for the one gene whose half life = 490min)
20 technical reps = 20 simulations, all simulations used dc_0_r_0 as input parameters

## Build S.cer.genom
```{r}
# Uses the genome for YAL010C, length = 493 codons
genomeString <- "29 23 33 58 29 6 36 52 23 39 0 11 58 37 59 49 17 56 60 50 36 31 59 57 9 7 18 50 1 48 46 38 48 27 27 7 10 40 20 33 44 0 18 16 25 36 20 45 30 21 47 51 33 31 48 10 31 47 27 7 11 47 50 39 45 39 18 31 15 47 25 60 58 27 57 45 7 0 37 36 28 9 21 10 29 41 30 47 51 7 19 32 27 36 7 1 49 8 48 57 38 36 28 36 32 30 24 31 10 60 55 60 60 2 31 50 28 60 60 6 30 49 48 53 6 31 6 22 21 27 23 17 6 46 21 11 55 21 21 45 26 58 58 15 38 29 57 57 33 59 47 7 27 8 0 29 18 18 21 40 23 60 32 36 49 36 11 29 26 22 15 53 60 60 10 21 8 59 27 30 55 27 50 4 58 11 36 38 7 47 16 41 31 27 37 9 56 18 11 45 49 60 7 23 27 5 15 58 38 52 27 16 31 10 26 49 50 35 45 22 11 30 49 44 25 57 31 31 47 46 28 46 26 15 3 8 11 56 27 14 27 52 60 27 59 33 15 5 46 48 51 27 38 58 57 48 17 47 48 30 48 12 40 32 23 51 28 48 27 47 56 31 32 27 10 13 17 18 45 45 48 58 46 1 22 48 14 48 31 47 51 11 4 2 22 58 7 11 31 26 58 46 20 8 44 31 26 44 11 14 4 8 11 56 36 21 22 17 17 28 26 8 49 31 21 30 31 31 7 21 27 8 32 19 45 6 8 28 55 7 18 31 32 30 59 38 2 51 21 23 25 16 8 31 52 34 7 25 31 44 3 55 30 7 20 35 47 48 23 7 18 35 55 16 21 36 22 23 27 31 7 27 51 58 0 10 46 46 44 27 38 21 19 7 8 8 38 47 49 19 8 21 11 7 30 21 18 31 60 45 20 11 49 60 55 56 21 27 59 50 44 27 43 6 22 51 27 21 23 27 56 8 13 21 56 39 12 11 27 18 47 1 14 48 9 25 52 10 51 38 13 11 36 8 60 27 45 7 7 8 22 31 7 31 0 18 47 18 44 0 51 7 48 8 30 13 31 18 32 55 10 34 0 22 11 13 18 36 10 37 57 45 48"

write.table(rep(genomeString,100), file = "../input/S.cer.genom", quote = FALSE, col.names = FALSE, row.names = FALSE)

```

## Build "allGenesDecrEqualsSynr_dc_0_r_0_S.cer.mRNA.ini.abndc.syn.dec"
```{r}
# tmp.syn.dec was built by: head -100 allGenesDecrEqualsSynrWithScaling_dc_0_r_0_S.cer.mRNA.ini.abndc.syn.dec.tsv > tmp.syn.dec
input.df <- read_tsv("../input/tmp.syn.dec", col_names = FALSE) 

input.df$X1 <- rep(0.0000138, 100)
input.df$X2 <- rep(200, 100)
input.df$X4 <- quantile(sort(simInputFeatures$mRNADecRateNeymotin_sec)[33:4839], probs = seq(0.01, 1, by = 0.01))
input.df$X3 <- input.df$X4 * 200

write_tsv(input.df, file = "../input/allGenesDecrEqualsSynr_dc_0_r_0_S.cer.mRNA.ini.abndc.syn.dec", col_names = FALSE)
```

## Build "amarel_troubleShoot_jobs.mf", should have 2 header lines + 3 * 20 jobs = 62 lines
```{r}
set.seed(200)  
# Seeds are required to feed each simulation. C program can set up to 32767 seeds.
seedvec <- sample(1:32767, 20, replace = F)  

maxTime <- 43260 # 60s burn in + 12h * 3600s/h = 43260s

jobFileAmr <- paste0("../ac_amarelCode/amarel_troubleShoot_jobs.mf") 
system(paste0("cp /dev/null ", jobFileAmr))  # To empty the job file from old content, otherwise contents will add up
write("BATCH_OPTIONS = -p p_shahlab_1,genetics_1,main -N 1 -n 1 -c 1 --mem=7G -t 04:00:00 -o stdout/%A.out -e error/%A.err -J troubleShoot-amr\n", file = jobFileAmr, append = T)

for (i in 1:20){
    outFileName = paste0("amr.troubleShoot.seed.", seedvec[i], "_dc_0_r_0_allGenesDecrEqualsSynr")
    write(paste0( ": ../input/S.cer.genom ../input/S.cer.tRNA ../input/allGenesDecrEqualsSynr_dc_0_r_0_S.cer.mRNA.ini.abndc.syn.dec ../ac_amarelCode/wrapper.sh ../source/SMOTNT\n", 
                  "    ./../ac_amarelCode/wrapper.sh -Tt ", maxTime, " -Tb 60 -R 200000 -t 3300000 -N 100 -dc 0 -r 0", " -F ../input/S.cer.genom -C ../input/S.cer.tRNA ", 
                  "-D ../input/allGenesDecrEqualsSynr_dc_0_r_0_S.cer.mRNA.ini.abndc.syn.dec -s ", seedvec[i], 
                  " -O ../output/", outFileName, 
                  " -p6 -p7\n"), file = jobFileAmr, append = T)
}
```

